src/video/VDP.hh

```diff
@@ -988,6 +988,11 @@ private:
 		void write(unsigned address, byte value, EmuTime::param time) override;
 	} vramPointerDebug;
 
+	struct LatchStatusDebug final : SimpleDebuggable {
+		explicit LatchStatusDebug(VDP& vdp);
+		[[nodiscard]] byte read(unsigned address) override;
+	} latchStatusDebug;
```

Comment:
I support the intend of this patch. Though I'm not sure if this is the right approach.

This debuggable exposes a single bit of information: whether there currently is a byte latched or not.
But then the next question will be: what is the value of this latched byte?

We could make a second debuggable to query this value. Or we could extend the size of this debuggable to two and expose this byte as the second position. Though that seems arbitrary: does the 1-bit-latch or the 8-bit-value go in position 0 or 1?

Ah, and next to `registerDataStored` there is also `paletteDataStored` (*1). Maybe we should then also expose this in the debuggable? Maybe the latched value in position 0, register-latched in position 1 and palette-latched in position 2?

(*1) Now that I look at it this from this new perspective: could it be that the real HW only has one latch-status for both `registerDataStored` and `palletteDataStored`? We should verify this on real HW. And if it turns out the real VDP only has a single latched-status, then we have to break the interface of our debuggable. That's the danger of exposing detailed information, maybe too detailed. Is the MSX programmer really interested in this?

Reply:
The use case for a real MSX programmer is described in this feature request:

https://github.com/openMSX/debugger/issues/148

Reply:
The openMSX implementation for those flags seems to be correct, see attached test. Both openMSX and real MSX2 give me a yellow background with an @ symbol about mid-way on the first line of the screen.

[vdplatch.zip](https://github.com/openMSX/openMSX/files/10572389/vdplatch.zip)

Reply:
@m9710797 , it might not be the optimal approach, but also it’s not that bad for a first implementation, is it?

At least the MSX programmer would have the feature.

Thinking in agile terms, it can be refined in more interactions for a more ideal approach.

Reply:
Sure, I agree it's a good idea.

But we should also be careful with breaking backwards compatibility in the debuggables. So if we can think/discuss a few days and then maybe have a better solution, than that's worth it IMHO.

Also because at the moment we don't have a good mechanism to detect the exact interface in the debugger and then adapt to it (if it would change between different openMSX versions).

But grauw's test has put me at easy. It's less likely we'll need to change it later.
Thanks grauw!

@pvmm: this discussion triggers another question/comment:
I assume one of your next patches will be to use this new debuggable in the debugger. Make sure that the debugger still works (e.g. doesn't crash) when combined with an older openMSX version that doesn't offer this new debuggable yet.

Reply:
> We could make a second debuggable to query this value. Or we could extend the size of this debuggable to two and expose this byte as the second position. Though that seems arbitrary: does the 1-bit-latch or the 8-bit-value go in position 0 or 1?

I don't think that is a problem provided we define a standard for latch statuses and stick to it. On that note, I think it makes more sense to put the status of the latch and then the value of the latch, since it is in the name of the debuggable itself, it should be the first (more important) piece of data.

On second thought, when debugging we know what's in the latch, since we can see the `out`s on the code view widget. I think it may be redundant.

> Ah, and next to `registerDataStored` there is also `paletteDataStored` (*1). Maybe we should then also expose this in the debuggable? Maybe the latched value in position 0, register-latched in position 1 and palette-latched in position 2?

Maybe a `{v99x8 latch statuses}` debuggable containing the two latches.

Reply:
> I don't think that is a problem provided we define a standard for debuggables of latch statuses and stick to it. On that note, I think it makes more sense to put the status of the latch and then the value of the latch, since it is in the name of the debuggable itself, it should be the first (more important) piece of data.

Ok. There are many designs that make sense. My main question was: "Have you already thought about this?" I'd like to avoid that later we extend this (e.g. with the palette stuff) but then it doesn't fit nicely anymore. For example if we picked the order 0->register-latch-status, 1->latch-value. Then later we can't cleanly extend it anymore with 2->palette-latch-status ... "Why is the order inconsistent: one 'status' in front and the other behind the 'value'?"

In other words, I'd like to see the larger design first before approving a small part of it. "See" either via code or via a textual description. And then it's fine if you don't yet fully implement it. But I want to avoid that we're painting ourselves into a corner. (And of course we can't predict everything we might want to change in the future, but we do already know about two possible extension on top of the current patch).
 
> On second thought, when debugging we know what's in the latch, since we can see the `out`s on the code view widget. I think it may be redundant.

Well, you could say the same about the latch-status.
Typically you're debugging exactly because something has gone wrong. And then maybe the OUT that wrote the latched-value is not easily visible (e.g. because the two OUTs are not cleanly paired anymore).

I don't feel strongly about whether this value should be shown in the debugger (by default) or not. But if we're extending openMSX with exposing this info, then IMHO we should try to be complete. In other words: it's fine for me that the debugger does not support visualizing everything that openMSX exposes.
 
> Maybe a `{VDP latch statuses}` debuggable containing the two latches. In the two LSBs of the char?

That's possible. And fine for me if you prefer that.
On the other hand, there's no advantage in cramming bits together (a larger debuggable doesn't cost anything more than a smaller debuggable). And using this information (e.g. from a Tcl script) might be more convenient if you can directly query (only) the stuff you're interested in, instead of also having to mask the desired bit. But I don't feel strongly about this, pick what you prefer.

> It starts by probing the contents of `debug list` for the existence of `{VDP latch status}`.

OK.

Reply:
There is a single `dataLatch` for all latch statuses. Would it be redundant to put the same data in the second byte of both {VDP latch status} and {Palette latch status}? I am leaning in favour of a third debugger for the latch value alone.

Reply:
All fine for me. I don't mind repeating the same latched-value in both debuggables.